src/topology/builder.rs

```diff
@@ -170,43 +170,95 @@ pub async fn build_pieces(
             .unwrap();
-        let input_rx = crate::utilization::wrap(Pin::new(input_rx));
+        let mut input_rx = crate::utilization::wrap(Pin::new(input_rx));
 
-        let (output, control) = Fanout::new();
+        let task = match transform {
+            Transform::Function(mut t) => {
+                let (output, control) = Fanout::new();
+
+                let transform = input_rx
+                    .filter(move |event| ready(filter_event_type(event, input_type)))
+                    .inspect(|_| emit!(EventIn))
+                    .flat_map(move |v| {
+                        let mut buf = Vec::with_capacity(1);
+                        t.transform(&mut buf, v);
+                        emit!(EventOut { count: buf.len() });
+                        stream::iter(buf.into_iter()).map(Ok)
+                    })
+                    .forward(output)
+                    .boxed()
+                    .map_ok(|_| {
+                        debug!("Finished.");
+                        TaskOutput::Transform
+                    });
+
+                outputs.insert(key.clone(), control);
 
-        let transform = match transform {
-            Transform::Function(mut t) => input_rx
-                .filter(move |event| ready(filter_event_type(event, input_type)))
-                .inspect(|_| emit!(EventIn))
-                .flat_map(move |v| {
-                    let mut buf = Vec::with_capacity(1);
-                    t.transform(&mut buf, v);
-                    emit!(EventOut { count: buf.len() });
-                    stream::iter(buf.into_iter()).map(Ok)
-                })
-                .forward(output)
-                .boxed(),
+                Task::new(key.clone(), typetag, transform)
+            }
+            Transform::FallibleFunction(mut t) => {
+                let (mut output, control) = Fanout::new();
+                let (mut errors_output, errors_control) = Fanout::new();
+
+                let transform = async move {
+                    while let Some(event) = input_rx.next().await {
+                        if filter_event_type(&event, input_type) {
+                            emit!(EventIn);
+                        }
+
+                        let mut buf = Vec::with_capacity(1);
+                        let mut err_buf = Vec::with_capacity(1);
+
+                        t.transform(&mut buf, &mut err_buf, event);
+                        // TODO: account for error outputs separately?
+                        emit!(EventOut { count: buf.len() });
+
+                        for event in buf {
+                            output.feed(event).await.expect("unit error");
+                        }
+                        output.flush().await.expect("unit error");
+                        for event in err_buf {
+                            errors_output.feed(event).await.expect("unit error");
+                        }
+                        errors_output.flush().await.expect("unit error");
```

Comment:
Do we actually enforce attaching a downstream component to the error output of a fallible transform? Or allow errors to be silently discarded?

If the former, then wouldn't the above end up eventually filling up the channel and eventually blocking?

Reply:
From a UX perspective, I think we want to allow users to not need to handle the error'd events if they don't want to.

Reply:
I would agree, but I didn't want to pigeonhole Luke's response. :P

Reply:
> ... if it has no outputs it'll simply drop anything written to the channel.

I'm so looking forward to the day where we have enough info in the topology to drop these without crossing a channel.